include/swift/Basic/StableHasher.h

```diff
@@ -0,0 +1,228 @@
+//===--- StableHasher.h - Stable Hashing ------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// An implementation of a stable hashing for Swift.
+//
+// Derived from the reference implementation for SipHash 2-4:
+//   https://github.com/veorq/SipHash
+//
+// With inline buffering derived from the hash implementation in the Swift
+// Standard Library.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_BASIC_STABLEHASHER_H
+#define SWIFT_BASIC_STABLEHASHER_H
+
+#include "swift/Basic/StableHasher.h"
+#include "llvm/Support/Endian.h"
+#include <algorithm>
+
+namespace swift {
+
+/// A \c StableHasher is an implementation of a 128-bit stable hash - built for
+/// speed.
+///
+/// A "stable" hash in this context refers to the idea that the output of this
```

Comment:
Does LLVM already have a hash like this? Should this be done upstream?

Reply:
Clang has an implementation of SipHash hidden behind the ptrauth stuff that does not present the correct API surface. That's the most likely place to modify should we wish to upstream.

Reply:
You're more than welcome to extend the API surface.